pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,259 @@
+	return allErrs
+}
+
+func ValidateFlowSchemaSubject(subject *flowcontrol.Subject, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if len(subject.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	}
+	switch subject.Kind {
+	case flowcontrol.ServiceAccountKind:
+		if len(subject.Name) > 0 {
+			for _, msg := range validation.ValidateServiceAccountName(subject.Name, false) {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("name"), subject.Name, msg))
+			}
+		}
+		if len(subject.APIGroup) > 0 {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("apiGroup"), subject.APIGroup, []string{""}))
+		}
+		if len(subject.Namespace) > 0 {
+			for _, msg := range apimachineryvalidation.ValidateNamespaceName(subject.Namespace, false) {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), subject.Namespace, msg))
+			}
+		} else {
+			allErrs = append(allErrs, field.Required(fldPath.Child("namespace"), "must specify namespace for service account"))
+		}
+	case flowcontrol.UserKind:
+		// TODO(ericchiang): What other restrictions on user name are there?
+		if subject.APIGroup != flowcontrol.GroupName {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("apiGroup"), subject.APIGroup, []string{flowcontrol.GroupName}))
+		}
+		if len(subject.Namespace) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), subject.Namespace, "must not set namespace for user-kind subject"))
+		}
+	case flowcontrol.GroupKind:
+		// TODO(ericchiang): What other restrictions on group name are there?
+		if subject.APIGroup != flowcontrol.GroupName {
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("apiGroup"), subject.APIGroup, []string{flowcontrol.GroupName}))
+		}
+		if len(subject.Namespace) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), subject.Namespace, "must not set namespace for group-kind subject"))
+		}
+	default:
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("kind"), subject.Kind, supportedSubjectKinds.List()))
+	}
+	return allErrs
+}
+
+func ValidateFlowSchemaPolicyRule(rule *flowcontrol.PolicyRule, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if len(rule.Verbs) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("verbs"), "verbs must contain at least one value"))
+	}
+
+	if len(rule.NonResourceURLs) > 0 {
+		if len(rule.APIGroups) > 0 || len(rule.Resources) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("nonResourceURLs"), rule.NonResourceURLs, "rules cannot apply to both regular resources and non-resource URLs"))
+		}
+		if hasWildcard(rule.NonResourceURLs) && len(rule.NonResourceURLs) > 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("nonResourceURLs"), rule.NonResourceURLs, "if '*' is present, must not specify other non-resource URLs"))
+		}
```

Comment:
What about checking the verb in this case?  IIRC, HTTP places some constraints but does not have a fixed vocabulary.  Also, does the kube apiserver support only a fixed vocabulary?

Reply:
for now, i remember only `GET` and `POST` is supported for non-resource urls. but let's keep in consistent w/ rbac validation for the current stage

Reply:
Yes, it's not a big problem to under-validate this; the only down side is that admins will be able to write a rule that matches nothing.